server/src/test/java/org/apache/iotdb/db/tools/TsFileSelfCheckToolTest.java

```diff
@@ -0,0 +1,211 @@
+            tablet.reset();
+          }
+        }
+        // write Tablet to TsFile
+        if (tablet.rowSize != 0) {
+          tsFileWriter.write(tablet);
+          tablet.reset();
+        }
+      }
+    } catch (Exception e) {
+      throw new Exception("meet error in TsFileWrite with tablet", e);
+    }
+  }
+
+  @Test
+  public void tsFileSelfCheckToolCompleteTest() {
+    TsFileSelfCheckTool tool = new TsFileSelfCheckTool();
+    try {
+      tool.check(path, false);
+    } catch (IOException | TsFileStatisticsMistakesException e) {
+      fail(e.getMessage());
+    }
+  }
+
+  @Test
+  public void tsFileSelfCheckToolWithStatisticsModifiedTest() throws IOException {
+    Map<Long, Pair<Path, TimeseriesMetadata>> timeseriesMetadataMap =
+        new TsFileSelfCheckTool().getTimeseriesMetadataMap(path);
+    for (Map.Entry<Long, Pair<Path, TimeseriesMetadata>> entry : timeseriesMetadataMap.entrySet()) {
+      TimeseriesMetadata timeseriesMetadata = entry.getValue().right;
+      Long pos = entry.getKey();
+      LongStatistics statistics = (LongStatistics) timeseriesMetadata.getStatistics();
+      statistics.initializeStats(666, 1999999, 1000000, 1999999, 0);
+
+      RandomAccessFile raf = new RandomAccessFile(path, "rw");
+      ByteArrayOutputStream bo = new ByteArrayOutputStream();
+      int serialLength = ReadWriteIOUtils.write(timeseriesMetadata.getTimeSeriesMetadataType(), bo);
+      serialLength += ReadWriteIOUtils.writeVar(timeseriesMetadata.getMeasurementId(), bo);
+      serialLength += ReadWriteIOUtils.write(timeseriesMetadata.getTSDataType(), bo);
+      serialLength +=
+          ReadWriteForEncodingUtils.writeUnsignedVarInt(
+              timeseriesMetadata.getDataSizeOfChunkMetaDataList(), bo);
+      serialLength += statistics.serialize(bo);
+      System.out.println("serialLength: " + serialLength);
+      byte[] serialArr = bo.toByteArray();
+      raf.seek(pos);
+      raf.write(serialArr, 0, serialArr.length);
+      raf.close();
+
+      break;
+    }
+
+    TsFileSelfCheckTool tool = new TsFileSelfCheckTool();
+    try {
+      tool.check(path, false);
+      fail("No exception thrown.");
+    } catch (TsFileStatisticsMistakesException e) {
+      assertEquals("Chunk exists statistics mistakes at position 22", e.getMessage());
+    }
+  }
+
+  @Test
+  public void tsFileSelfCheckToolWithRandomModifiedTest() throws IOException {
+
+    RandomAccessFile raf = new RandomAccessFile(path, "rw");
+    ByteArrayOutputStream bo = new ByteArrayOutputStream();
+    ReadWriteIOUtils.write(100, bo);
+    byte[] serialArr = bo.toByteArray();
+    // timeseriesMetadata begins at 878364
+    // randomly modify timeseriesMetadata region
+    raf.seek(878375);
+    raf.write(serialArr, 0, serialArr.length);
+    raf.close();
+
+    TsFileSelfCheckTool tool = new TsFileSelfCheckTool();
+    try {
+      tool.check(path, false);
+      fail("No exception thrown.");
+    } catch (BufferUnderflowException e) {
```

Comment:
May be we should seal this exception into a higher exception and give some error messages.